src/script/conversation/ConversationRepository.ts

```diff
@@ -310,13 +312,154 @@ export class ConversationRepository {
+      .conversations()
+      .filter(conversation => conversation.domain !== deletedDomain);
+
+    conversationsToRemoveTheirDeletedDomainUsers.forEach(async conversation => {
+      const usersToRemove = conversation.allUserEntities().filter(user => user.domain === deletedDomain);
+      if (usersToRemove.length === 0) {
+        return;
+      }
+      await this.insertFederationStopSystemMessage(conversation, [deletedDomain]);
+      await this.removeDeletedFederationUsers(conversation, usersToRemove);
+    });
+  };
+
+  /**
+   * For the `federation.connectionRemoved` event: (Backend A & B stopped federating, user is on C)
+    - receive the event from backend
+    - Identify all conversations that are not owned from A or B domain and that contain users from A and B
+      - remove users from A and B from those conversations
+      - insert system message in those conversations about backend A and B stopping to federate
+    - identify all conversations owned by domain A that contains users from B
+      - remove users from B from those conversations
+      - insert system message in those conversations about backend A and B stopping to federate
+    - Identify all conversations owned by domain B that contains users from A
+      - remove users from A from those conversations
+      - insert system message in those conversations about backend A and B stopping to federate
+   * @param domains The domains that stopped federating with each other
+   */
+  private readonly onFederationConnectionRemove = async (domains: string[]) => {
+    const selfUser = this.userState.self();
+    const [domainOne, domainTwo] = domains;
+    const allConversations = this.conversationState.conversations();
+
+    allConversations
+      .filter(conversation => conversation.domain === domainOne)
+      .forEach(async conversation => {
+        const usersToDelete = conversation.allUserEntities().filter(user => user.domain === domainTwo);
+        if (usersToDelete.length > 0) {
+          await this.removeDeletedFederationUsers(conversation, usersToDelete);
+          await this.insertFederationStopSystemMessage(conversation, [domainOne, domainTwo]);
+        }
+      });
+
+    allConversations
+      .filter(conversation => conversation.domain === domainTwo)
+      .forEach(async conversation => {
+        const usersToDelete = conversation.allUserEntities().filter(user => user.domain === domainOne);
+        if (usersToDelete.length > 0) {
+          await this.removeDeletedFederationUsers(conversation, usersToDelete);
+          await this.insertFederationStopSystemMessage(conversation, [domainOne, domainTwo]);
+        }
+      });
+
+    allConversations
+      .filter(conversation => {
+        if (conversation.domain !== selfUser.qualifiedId.domain) {
+          return false;
+        }
+
+        const hasAnyUserFromDomainOne = conversation
+          .allUserEntities()
+          .find(user => user.qualifiedId.domain === domainOne);
```

Comment:
Instead of using .find() twice in the filter, we can convert the list of user domains in a conversation into a Set for faster lookups, and then check for the existence of domainOne and domainTwo in the Set.



```
const userDomains = new Set(
      conversation.allUserEntities().map(user => user.qualifiedId.domain)
    );
```

    return userDomains.has(domainOne) && userDomains.has(domainTwo);

Reply:
nice https://github.com/wireapp/wire-webapp/pull/15583/commits/fc35485500af1f1fc508cf1a7a29b10e46460b8a